src/storage/src/hummock/event_handler/uploader.rs

```diff
@@ -445,44 +428,291 @@ impl SealedData {
+                    is unspilled data in previous epoch: prev epoch {}, spill epoch {}",
+                        oldest_sealed_epoch.epoch, epoch
+                    );
+                }
+                Ordering::Equal => {
+                    let epoch_data = self.sealed_data.pop_back().unwrap();
+                    assert_eq!(epoch, epoch_data.epoch);
+                    epoch_data.imms
+                }
+                Ordering::Greater => VecDeque::new(),
+            }
+        } else {
+            match self.current_epoch_data.epoch.cmp(&epoch) {
+                Ordering::Less => {
+                    assert!(
+                        self.current_epoch_data.imms.is_empty(),
+                        "should not spill at this epoch because there \
+                    is unspilled data in current epoch epoch {}, spill epoch {}",
+                        self.current_epoch_data.epoch,
+                        epoch
+                    );
+                    VecDeque::new()
+                }
+                Ordering::Equal => {
+                    if !self.current_epoch_data.imms.is_empty() {
+                        take(&mut self.current_epoch_data.imms)
+                    } else {
+                        VecDeque::new()
+                    }
+                }
+                Ordering::Greater => VecDeque::new(),
+            }
+        };
+        self.add_flushing_imm(imms.iter().rev().map(|imm| imm.batch_id()));
+        imms.into_iter().collect()
+    }
+
+    fn add_flushing_imm(&mut self, imm_ids: impl Iterator<Item = SharedBufferBatchId>) {
+        for imm_id in imm_ids {
+            if let Some(prev_imm_id) = self.flushing_imms.front() {
+                assert_gt!(imm_id, *prev_imm_id);
+            }
+            self.flushing_imms.push_front(imm_id);
+        }
+    }
+
+    // start syncing the imm inclusively before the `epoch`
+    // returning data with newer data coming first
+    fn sync(&mut self, epoch: HummockEpoch) -> Vec<ImmutableMemtable> {
+        assert_lt!(epoch, self.current_epoch_data.epoch());
+        // firstly added from old to new
+        let mut ret = Vec::new();
+        while let Some(epoch_data) = self.sealed_data.back()
+            && epoch_data.epoch() <= epoch
+        {
+            let imms = self.sealed_data.pop_back().expect("checked exist").imms;
+            self.add_flushing_imm(imms.iter().rev().map(|imm| imm.batch_id()));
+            ret.extend(imms.into_iter().rev());
+        }
+        // reverse so that newer data comes first
+        ret.reverse();
+        ret
+    }
+}
+
+#[derive(Default)]
+struct UnsyncData {
+    instance_data: HashMap<LocalInstanceId, LocalInstanceUnsyncData>,
+    // An index of `instance_data` that maintains the set of existing instance id of a table id
+    table_instance_id: HashMap<TableId, HashSet<LocalInstanceId>>,
+    epoch_data: BTreeMap<HummockEpoch, EpochData>,
```

Comment:
Need some documentation to explain the transistion between `instance_data` and `epoch_data`. IIUC, all data will go to `instance_data` first and move to `epoch_data` on sync/spill.